### ISSUE TITLE
Support various behaviors when increasing sample length after drawing a waveform in BitInvader

### ISSUE THREAD

**Issue by Sawuare** [LMMS/lmms#3797]
To re-produce:
1. Load a default BitInvader.
2. Decrease the sample length (e.g., to `64`).
3. Draw a waveform or click any waveform button (e.g., sine wave button).
4. Increase the sample length (e.g., to `128`).

The odd behavior:
![image](https://user-images.githubusercontent.com/25503477/30217720-b0926088-94bf-11e7-8d73-dfbfbe1dcc1a.png)

The expected behavior:
![capture20170908175912396](https://user-images.githubusercontent.com/25503477/30217881-275ad5d8-94c0-11e7-81b9-552df6b5aef2.png)

**Comment by PhysSong** [comment]
Introduced in #1362, from changes in `Graph.cpp`. After that PR, the waveform is preserved on resizing, but doesn't clear its invisible "tail" part on waveform change.
See [differences](https://github.com/LMMS/lmms/pull/1362/files#diff-117e2406b2d89518666fbc4b44a66e44R492) for details.

**Comment by PhysSong** [comment]
Before we fix this, we should discuss what is the desired behavior. I think the second picture is what the BitInvader supposed to behave. Since the second behavior was in 1.1 and there were no related bug reports as far as I know. And, current behavior is very weird. :-1:
However someone might think other waveform(such as two periods of sine wave) should be there. So we need to discuss.

**Comment by musikBear** [comment]
> someone might think other waveform(such as two periods of sine wave) should be there

@PhysSong -Yes, periodic repeats would imo be the solution to go for. The reason would be, that *perfect* periodic repeats are very difficult to draw manually, whereas a simple strait line (img 2) is very easy to make manually, so imo: let the computer do the difficult thing.

**Comment by Sawuare** [comment]
Periodically repeating a wave is just increasing its frequency/pitch...
IMO, this is useless, as there is already a pitch parameter

> perfect periodic repeats are very difficult to draw manually, whereas a simple strait line (img 2) is very easy to make manually

Drawing a straight line at a level of zero (silence) is hard.

IMO, adding silence when increasing sample length is the best behavior.

**Comment by DouglasDGI** [comment]
So the final verdict is to add silence when the sample length is increased?  I can get this done if you would like me to, and combine the changes with those added in #4705 as well.

**Comment by DouglasDGI** [comment]
In #4705 I made it so changing the sample length changes the sound realtime.  I think the current graph behavior should be preserved, so that users may be able to move the sample length knob to change their sound in interesting ways.

If we instead make it so anything off the edge of the graph is changed to silence, then the graph would have to be redrawn whenever the sample length knob is automated, which I don't think is the wanted behavior.  It would also render #4705 pointless.

One or the other would have to be chosen, and I think #4705 would be the wanted change, since #4705 adds sound design capabilities and #3797 removes some slightly.  What do you all think?

**Comment by zonkmachine** [comment]
> If we instead make it so anything off the edge of the graph is changed to silence, then the graph would have to be redrawn whenever the sample length knob is automated, which I don't think is the wanted behavior. It would also render #4705 pointless.

Can't we add a button to toggle the 'clearing of the end' on/off?

**Comment by DomClark** [comment]
Perhaps we could keep the current behaviour and add a "flat line" waveform, for a more consistent way to achieve both results.

**Comment by PhysSong** [comment]
Suppose I'm doing this:
1. Increase the length to 192
2. Set the waveform to sine wave
3. Decrease the length to 128
4. Set the waveform to sine wave
5. Increase the length to 192 again

Then I can think of those three behaviors:

|1.2.0-RC7|1.1.3|Alternative|
|---|---|---|
|![12rc](https://user-images.githubusercontent.com/12591074/51228129-8e9fb500-199a-11e9-9393-4d007099943e.png)|![113](https://user-images.githubusercontent.com/12591074/51228130-8f384b80-199a-11e9-8015-912ab72d0f60.png)|![extend](https://user-images.githubusercontent.com/12591074/51228131-8f384b80-199a-11e9-9fc2-56253094bfa9.png)|

**Comment by PhysSong** [comment]
> If we instead make it so anything off the edge of the graph is changed to silence, then the graph would have to be redrawn whenever the sample length knob is automated, which I don't think is the wanted behavior.

@DouglasDGI It doesn't affect changing the length of the graph. Instead, it affects how the 'invisible' part changes when selecting a waveform.
Also, the graph will be redrawn when automating the length, regardless of the tail's behavior.

**Comment by DouglasDGI** [comment]
Whatever the final decision is, I'd probably be willing to code it.  It shouldn't be too difficult.

**Comment by jasp00** [comment]
If the three behaviors PhysSong has represented are desired, all of them should be available as a "resize behavior" input.

**Comment by DouglasDGI** [comment]
@jasp00 
> extra points for coding upgrade steps.
What do you mean by this?

Also, instead of a button I could use a simple dropdown box (like the one in AudioFileProcessor) to choose between the modes, so there isn't any hassle if we want more than just two behaviors implemented in the future.

**Comment by jasp00** [comment]
>What do you mean by this?

Regarding upgrade steps, `src/core/DataFile.cpp` is in charge of upgrading old projects. You should add an upgrade step for projects before 1.2.0-rc7 and another one for 1.2.0-rc7.

>instead of a button I could use a simple dropdown box

With two options only, a button would be less intrusive. However, use your judgment.

**Comment by PhysSong** [comment]
@DouglasDGI Do you still want to work on this? For 1.2, I think restoring 1.1's behavior is fine. Then we need to change `graphModel::setWaveToSine` and its siblings.
Also, I bumped this issue to 1.2.0. If this one blocks releasing 1.2.0, please feel free to bump it to 1.2.1 or later.

**Comment by DouglasDGI** [comment]
@PhysSong I'll see if I can finish it today.

> Then we need to change `graphModel::setWaveToSine` and its siblings.

I'd rather avoid doing that, as that would *require* a loss of off-graph information for any synthesizers that need to keep the off-graph information (including Microwave actually, that synthesizer I'm programming).  

Instead, I'll add `graphModel::clearAll`, which clears all samples regardless of whether they're visible.  Then I'll put that in the functions BitInvader runs when a waveform button is clicked.  I just did that and tested it out, and it works just fine.  Would that be alright?

**Comment by DouglasDGI** [comment]
If that's okay then I can probably go ahead and make a PR right now.

**Comment by PhysSong** [comment]
The original issue is fixed via 8d707df171fa89c8426dca2d0d42de744d5db365. Now changing this issue as an enhancement for 1.3.0.

**Comment by zonkmachine** [comment]
This was closed in https://github.com/LMMS/lmms/pull/4824 and https://github.com/LMMS/lmms/issues/5799 which restores the behavior from 1.1.3 .

The default sine wave that is created when a new wave is initialized fills the whole sample. When you generate/draw a wave that is shorter than this, the rest of the sample was unchanged. Its doesn't seem practical to make an option to leave this in nor does it seem very useful musically. There are wave table editors out there that have more advanced features. The best I saw was the standalone editor for wiard waveform city synth module http://www.wiard.com/modular/300series/waveformcity/index.htm but unfortunately it doesn't seem to be available any longer.
